refactor(scan): implement CI discovery provider

ostridm opened the pull request:
closes #88

ostridm committed:
feat(scan): implement CI discovery provider

closes #88

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -1,9 +1,48 @@
+      .Where(x => x.Name.Equals(vendor.Constant, StringComparison.OrdinalIgnoreCase))
+      .Select(x => x.GetValue(null))
+      .Cast<CiServer>()
+      .FirstOrDefault() ?? new CiServer(vendor.Name);
```
`CiServer` is an enumeration, why don't move that logic to it? (e.g. `FromValue()` or `FromName`)

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -2,5 +2,50 @@ namespace SecTester.Scan.CI;
 
 public record CiServer(string? ServerName = default)
```
It is hardly a record, covert to class and implement a `IComparable` interface

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -2,5 +2,50 @@ namespace SecTester.Scan.CI;
 public record CiServer(string? ServerName = default)
 {
-  public static readonly CiServer? Unknown;
+  public static CiServer APPCENTER { get; } = new("Visual Studio App Center");
```
Please follow the naming convention, instead of using a property name introduce an `id` or `value` field. Use it to filter enumeration.

```suggestion
  public static CiServer AppCenter { get; } = new("Visual Studio App Center");
```

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -1,9 +1,48 @@
+    env ??= Environment.GetEnvironmentVariables();
+
+    var vendors = JsonSerializer.Deserialize<Vendor[]>(
+      ResourceUtils.GetEmbeddedResourceContent<DefaultCiDiscovery>("SecTester.Scan.CI.vendors.json"),
```
Extract a string as constant

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+      x.Value.ValueEquals(_env[x.Name]?.ToString()));
+  }
+
+  private bool MatchInnerAny(JsonElement element)
```
```suggestion
  private bool Any(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+      .Any(x => x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!));
+  }
+
+  private bool MatchInnerEnvIncludes(JsonElement element)
```
```suggestion
  private bool Includes(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+    return envVarValue?.Contains(includes) ?? false;
+  }
+
+  private bool MatchInnerEnvNe(JsonElement element)
```
```suggestion
  private bool NotEqual(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+    };
+  }
+
+  private bool MatchOwnProperties(JsonElement element)
```
```suggestion
  private bool HasOwnProperties(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+    _env = env;
+  }
+
+  public bool MatchEnvElement(JsonElement element)
```
```suggestion
  public bool MatchEnv(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+  }
+
+
+  public bool MatchPrElement(JsonElement element)
```
```suggestion
  public bool MatchPr(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+    return !string.IsNullOrEmpty(envVarValue);
+  }
+
+  private static string? GetInnerPropertyValue(JsonElement element, string name)
```
```suggestion
  private static string? GetPropertyValue(JsonElement element, string name)
```

derevnjuk commented on test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs:
```diff
@@ -0,0 +1,252 @@
+  {
+    new object[]
+    {
+      @"{
+        ""env"": { ""env"": ""NODE"", ""includes"": ""/app/.heroku/node/bin/node""}
+        }",
```
Avoid the wrong indentation
```suggestion
      @"{""env"": { ""env"": ""NODE"", ""includes"": ""/app/.heroku/node/bin/node""}}",
```

derevnjuk commented on test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs:
```diff
@@ -0,0 +1,252 @@
+using System.Collections.Specialized;
```
global using

derevnjuk commented on test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs:
```diff
@@ -0,0 +1,252 @@
+    //assert
+    result.Should().Be(expected);
+  }
+
+  [Fact]
+  public void MatchEnvElement_GivenAnyEnv_Returns()
+  {
+    // arrange
+    var sut = new VendorMatcher(new Dictionary<string, object>());
+    var vendors = JsonSerializer.Deserialize<Vendor[]>(
+      ResourceUtils.GetEmbeddedResourceContent<DefaultCiDiscovery>("SecTester.Scan.CI.vendors.json"),
+      new JsonSerializerOptions() { PropertyNameCaseInsensitive = true });
+
+    // act
+    var result = vendors!.FirstOrDefault(x=> sut.MatchEnvElement(x.Env));
+    
+    // assert
+    result.Should().BeNull();
+  }
+  
+  [Fact]
+  public void MatchPrElement_GivenAnyPr_Returns()
+  {
+    // arrange
+    var sut = new VendorMatcher(new Dictionary<string, object>());
+    var vendors = JsonSerializer.Deserialize<Vendor[]>(
+      ResourceUtils.GetEmbeddedResourceContent<DefaultCiDiscovery>("SecTester.Scan.CI.vendors.json"),
+      new JsonSerializerOptions() { PropertyNameCaseInsensitive = true });
+
+    // act
+    var result = vendors!.FirstOrDefault(x=> sut.MatchPrElement(x.Pr));
+    
+    // assert
+    result.Should().BeNull();
+  }
```
Those tests do not test anything, they just overlap the tests above

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+  {
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => element.GetString() is not null && _env.Contains(element.GetString()!),
```
It may lead to FP in the case of an empty string. Also, you have to pass `StringComparison` second argument

derevnjuk replied:
Also duplicated multiple times

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,88 @@
+      JsonValueKind.String => element.GetString() is not null && _env.Contains(element.GetString()!),
+      JsonValueKind.Object => MatchInnerAny(element) || MatchInnerEnvIncludes(element) ||
+                              MatchOwnProperties(element),
+      JsonValueKind.Array => element.EnumerateArray()
+        .All(x => x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)),
```
Duplicate `MatchInnerAny`

ostridm replied:
`All` vs `Any`

derevnjuk replied:
What difference that makes?

```diff
- x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)
+ x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)
```

derevnjuk reviewed (changes requested):
Look fine, but you still have what needs to be improved and refactored

ostridm committed:
feat(scan): implement CI discovery provider

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044265604
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044268563
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044272236
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044272236
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044273222
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044273222
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044281812
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044281957
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044282974
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044283646
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044284132
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044284252
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044285109
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044286325
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044288520
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044293428
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044294429
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1044295947

closes #88

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+  private readonly string _name;
+  private readonly int _hashcode;
+
+  public string Constant => _constant;
```
`Constant`? What does it mean? You must utilize clear and concise names, e.g. `Name` and `DisplayName`, or `Id` and `Name`.

derevnjuk replied:
ditto

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+      return false;
+    }
+
+    return ReferenceEquals(Constant, other.Constant) || Constant.Equals(other.Constant, StringComparison.OrdinalIgnoreCase);
```
It is not clear why you need to do the same by `ReferenceEquals`

```suggestion
    return Constant.Equals(other.Constant, StringComparison.OrdinalIgnoreCase);
```

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+  public override string ToString()
+  {
+    return _constant;
+  }
```
It is not clear to me why don't show humanized branded name

```suggestion
  public override string ToString() => _name;
```

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+  public override int GetHashCode()
+  {
+    return _hashcode;
+  }
```
```suggestion
  public override int GetHashCode() => _hashcode;
```

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+  {
+    return From(vendor.Constant, vendor.Name);
+  }
+  public static CiServer From(string constant, string name)
```
```suggestion
  }
  
  public static CiServer From(string constant, string name)
```

derevnjuk replied:
Well, quite a strange method 🤦🏾  In your interpretation this is a static factory method, it seems you love to utilize patterns randomly)

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+    return typeof(CiServer)
+             .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+             .Select(x => x.GetValue(null))
+             .Cast<CiServer>()
```
Extract to a separate method (e.g.  `GetAll`)

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+             .Select(x => x.GetValue(null))
+             .Cast<CiServer>()
+             .FirstOrDefault(x => x.Constant.Equals(constant, StringComparison.OrdinalIgnoreCase))
+           ?? new CiServer(constant, name);
```
Return null instead, deciding what to do is the responsibility of the caller code

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,125 @@
+  private readonly int _hashcode;
+
+  public string Constant => _constant;
+  public string Name => _name;
```
Why don't convert it to auto-prop?

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,89 @@
+    };
+  }
+
+  private bool CheckStringValue(JsonElement element)
```
Unclear and superficial name, it is intended to check what? (e.g. `HasProperty`)

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,89 @@
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => CheckStringValue(element),
+      JsonValueKind.Object => NotEqual(element) || MatchEnv(element),
```
What is the point to call the whole `MatchEnv`? Extract `JsonValueKind.Object => Any(element) || Includes(element) || HasOwnProperties(element)` to a separate method

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,89 @@
+    };
+  }
+
+  public bool MatchPr(JsonElement element)
```
The only difference between `MatchEnv` and `MatchPr` is in using `NotEqual`. Why don't extract a common logic to a separate method

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,89 @@
+    return any.Value.ValueKind == JsonValueKind.Array && any.Value.EnumerateArray().Any(CheckStringValue);
+  }
+
+  private bool Includes(JsonElement element)
```
`Includes` and `NotEqual` are identical, the difference only in the last statement
```csharp
return envVarValue?.Contains(includes) ?? false;
// vs
return ne == "false" && !string.IsNullOrEmpty(envVarValue);
```
or even better
```csharp
return envVarValue?.Contains(includes) ?? false;
// vs
return !envVarValue?.Equals(ne) ?? false;
```

ostridm committed:
refactor(scan): implement CI discovery provider

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045191678
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045192284
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045192330
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045192650
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045193119
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045193362
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045192100
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045194634
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045194641
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045195231
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045192937
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045196975
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045197819
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045198688

closes #88

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,107 @@
+  private static readonly IEnumerable<CiServer> All = typeof(CiServer)
+    .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+    .Select(x => x.GetValue(null))
+    .Cast<CiServer>();
```
Avoid any redundant optimization made without any purpose, especially when it was not requested. Doing any useless changes after review, which have not been requested, just delays the merge. Move to the method.

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,107 @@
+  {
+    if (string.IsNullOrEmpty(id))
+    {
+      throw new ArgumentException("Id must not be empty.");
```
Can it accept null? Clarify the message or use `ArgumentNullException`

```suggestion
      throw new ArgumentException("ID must not be empty.");
```

derevnjuk commented on src/SecTester.Scan/CI/CiServer.cs:
```diff
@@ -1,6 +1,107 @@
+  public override bool Equals(object? obj)
+  {
+    return Equals(obj as CiServer);
+  }
```
```suggestion
  public override bool Equals(object? obj) => Equals(obj as CiServer)
```

derevnjuk commented on src/SecTester.Scan/CI/DefaultCiDiscovery.cs:
```diff
@@ -1,9 +1,44 @@
+    }
+
+    Server = CiServer.GetAll()
+               .FirstOrDefault(x => vendor.Constant.Equals(x.Id, StringComparison.OrdinalIgnoreCase))
```
Why did you remove `FromId` (aka `From`)? If you would be a hero, your name was "Duplicatemen" 😄  You make me repeat comments multiple times https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045193362

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,96 @@
+    return Any(element) || Includes(element) || HasOwnProperties(element);
+  }
+
+  private bool HasValue(JsonElement element)
```
`Contains` just verifies if there is a property, but cannot say anything about value https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045195231

```suggestion
  private bool HasProperty(JsonElement element)
```

derevnjuk commented on src/SecTester.Scan/CI/VendorMatcher.cs:
```diff
@@ -0,0 +1,96 @@
+
+  private bool ApplyPredicate(JsonElement element, string propertyName, Func<string?, string, bool> predicate)
+  {
+    var env = GetPropertyValue(element, "env");
```
The method is too complex. Extract a logic to retrieve env to a separate method.

```csharp
private string? GetEnvValue(JsonElement element)
{
  var env = GetPropertyValue(element, "env");

  if (string.IsNullOrEmpty(env) || !_environment.Contains(env))
  {
    return null;
  }

  return _environment[env]?.ToString();
}
```

derevnjuk commented on test/SecTester.Scan.Tests/CI/CiServerTests.cs:
```diff
@@ -0,0 +1,100 @@
+    }
+  };
+
+  public static IEnumerable<object[]> EqualityInput = new List<object[]>
```
OMG... Never use functions to generate test data in suites, such tests are not possible to maintain at all. I had already rewritten the whole scan tests because of this, changing just a line of code

derevnjuk commented on test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs:
```diff
@@ -0,0 +1,179 @@
+
+  [Theory]
+  [MemberData(nameof(MatchPrInput))]
+  public void MatchPrElement_ReturnsExpected(string vendorInput, IDictionary environmentInput, bool expected)
```
The test name and method are different

derevnjuk commented on test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs:
```diff
@@ -0,0 +1,179 @@
+
+  [Theory]
+  [MemberData(nameof(MatchEnvInput))]
+  public void MatchEnvElement_ReturnsExpected(string vendorInput, IDictionary environmentInput, bool expected)
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs:
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
+
     // assert
-    _sut.IsCi.Should().BeFalse();
+    sut.Server.Should().BeSameAs(CiServer.GithubActions);
+    sut.IsCi.Should().BeTrue();
+    sut.IsPr.Should().BeTrue();
```
You must use the partial match, we have already discussed this multiple times

derevnjuk commented on test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs:
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
+
     // assert
-    _sut.Server.Should().BeNull();
+    sut.Server.Should().BeSameAs(CiServer.GithubActions);
+    sut.IsCi.Should().BeTrue();
+    sut.IsPr.Should().BeFalse();
```
ditto

derevnjuk commented on test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs:
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
+    var sut = new DefaultCiDiscovery(new Dictionary<string, string>());
+
+    // assert
+    sut.Server.Should().BeNull();
+    sut.IsCi.Should().BeFalse();
+    sut.IsPr.Should().BeFalse();
```
ditto

ostridm committed:
refactor(scan): implement CI discovery provider

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045840702
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045842613
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045843292
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045844983
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045851116
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045867250
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045873232
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045873295
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045874963
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045875082
https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045875389

closes #88
